zcash_client_sqlite/src/lib.rs

```diff
@@ -161,30 +162,48 @@ pub(crate) const UA_TRANSPARENT: bool = true;
 pub(crate) const DEFAULT_UA_REQUEST: UnifiedAddressRequest =
     UnifiedAddressRequest::unsafe_new(UA_ORCHARD, true, UA_TRANSPARENT);
 
-/// The ID type for accounts.
+/// Unique identifier for a specific account tracked by a [`WalletDb`].
+///
+/// Account identifiers are "one-way stable": a given identifier always points to a
+/// specific viewing key within a specific [`WalletDb`] instance, but the same viewing key
+/// may have multiple account identifiers over time. In particular, this crate upholds the
+/// following properties:
+///
+/// - When an account starts being tracked within a [`WalletDb`] instance (via APIs like
+///   [`WalletWrite::create_account`], [`WalletWrite::import_account_hd`], or
+///   [`WalletWrite::import_account_ufvk`]), a new `AccountUuid` is generated.
+/// - If an `AccountUuid` is present within a [`WalletDb`], it always points to the same
+///   account.
+///
+/// What this means is that account identifiers are not stable across "wallet recreation
+/// events". Examples of these include:
+/// - Restoring a wallet from a backed-up seed.
+/// - Importing the same viewing key into two different wallet instances.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Default)]
-pub struct AccountId(u32);
+pub struct AccountUuid(Uuid);
 
-impl AccountId {
-    /// Constructs an `AccountId` from a bare `u32` value. The resulting identifier is not
-    /// guaranteed to correspond to any account stored in the database.
-    #[cfg(feature = "unstable")]
-    pub fn from_u32(value: u32) -> Self {
-        AccountId(value)
+impl AccountUuid {
+    /// Constructs an `AccountUuid` from a bare [`Uuid`] value.
+    ///
+    /// The resulting identifier is not guaranteed to correspond to any account stored in
+    /// a [`WalletDb`]. Callers should use [`WalletDb::get_account_for_uuid`]
+    pub fn from_uuid(value: Uuid) -> Self {
+        AccountUuid(value)
     }
 
-    /// Unwraps a raw `accounts` table primary key value from its typesafe wrapper.
-    ///
-    /// Note that account identifiers are not guaranteed to be stable; if a wallet is restored from
-    /// seed, the account identifiers of the restored wallet are not likely to correspond to the
-    /// identifiers for the same accounts in another wallet created or restored from the same seed.
-    /// These unwrapped identifier values should therefore be treated as ephemeral.
-    #[cfg(feature = "unstable")]
-    pub fn as_u32(&self) -> u32 {
+    /// Exposes the opaque account identifier from its typesafe wrapper.
+    pub fn expose_uuid(&self) -> Uuid {
         self.0
     }
 }
 
+/// The local identifier for an account.
+///
+/// This is an ephemeral value for efficiently and generically working with accounts in a
+/// [`WalletDb`]. To reference accounts in external contexts, use [`AccountUuid`].
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Default)]
+pub struct AccountId(u32);
```

Comment:
Would it be useful to rename this to `AccountLocalId`? Doing so is obviously a more disruptive change, but it would force all code currently using `AccountId` to explicitly choose between `AccountUuid` and `AccountLocalId`. It would also, I think, reduce the chance of confusing these with ZIP 32 account indices.

Reply:
Such a rename serves little purpose, because it is named directly for the `WalletRead::AccountId` associated type that it gets set to, and the associated type is what actually gets used in most places.

I'm somewhat annoyed that we now have this split where our local account ID that needs to be in `WalletRead` is infallible concretely but fallible in the trait API, which was in part due to #1235 getting rushed into a crate release (something I expressed disagreement with at the time). As it is, we'd need Yet Another Large Refactor to fix it, and I explicitly do not want to do that at the same time as these changes; our changelogs get large enough as it is.

Reply:
Confusion with the associated type is one reason I want to do this change, actually. It makes no sense to me to have the associated type be named the same as *one particular* possible instantiation of it. They are conceptually not the same thing.

Reply:
I have a different take here: `WalletRead::AccountId` should now be `AccountUuid`, and `AccountId` should be removed. The whole point is that the identifier is that there is a single, existentially typed identifier for the account. By adding `AccountUuid` instead of modifying `AccountId` to contain a UUID, you've created a confusing situation where there are two different ways to reference an account.

Reply:
My preference is for the transition to instantiate `WalletRead::AccountId` as `AccountUuid` be part of this PR; that then obviates the need for a separate method to query for account data by UUID, and we don't really want anyone writing code that targets the intermediate API state that you're talking about.

Reply:
@nuttycom Originally my plan _was_ to change `AccountId` to be the UUID-based thing (https://github.com/zcash/librustzcash/issues/1335#issuecomment-2477464017). But then you said the exact opposite of the above in https://github.com/zcash/librustzcash/issues/1335#issuecomment-2477484782: use two types. So that's what I did here.

Reply:
> > Augment WalletRead::AccountId with the ability to encode to, and parse from, opaque bytes. This encoding MUST NOT be parsed or otherwise interpreted by wallet apps.
>
> This doesn't need to be something that's exposed at the `zcash_client_backend` level, because the user of `zcash_client_sqlite` can use the knowledge that they're depending on that crate; they don't need this to be generic over data storage implementations.

I think this might be a miscommunication? I interpret it only as saying that the *serialization* shouldn't be exposed by `zcash_client_backend`. (FWIW I disagree and think that `WalletRead::AccountId` should be generically serializable, but it's only mild disagreement, because you should never mix serializations of it from different backends.)

> > Encode its account IDs (which are currently the account's primary key from the accounts table) as the tuple (wallet_uuid, account_primary_key).
>
> An alternative that we discussed to this was to assign a distinct UUID to each account, and map that to an account ID prior to performing queries. After some consideration, I think that I prefer this solution to the tuple-based approach, because when we add a persistent database reference, we can just have an in-memory map that acts as a cache of the `(account_uuid, account_primary_key)` mappings.

As I interpret this, it's primarily talking about the decision to use `AccountUuid` rather than `(WalletUuid, LocalAccountId)`. Using an in-memory map was secondary and was only one possibility.